File: Klooni1010/core/src/dev/lonami/klooni/effects/WaterdropEffectFactory.java
package dev.lonami.klooni.effects;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.g2d.Batch;
import com.badlogic.gdx.math.MathUtils;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.math.Vector3;
import dev.lonami.klooni.SkinLoader;
import dev.lonami.klooni.game.Cell;
import dev.lonami.klooni.interfaces.IEffect;
import dev.lonami.klooni.interfaces.IEffectFactory;


public class WaterdropEffectFactory implements IEffectFactory {
    private Texture dropTexture;


    private void init() {
        if (dropTexture == null)
            dropTexture = SkinLoader.loadPng("cells/drop.png");
    }

    @Override
    public String getName() {
        return "waterdrop";
    }

    @Override
    public String getDisplay() {
        return "Waterdrop";
    }

    @Override
    public int getPrice() {
        return 200;
    }

    @Override
    public IEffect create(Cell deadCell, Vector2 culprit) {
        init();
        IEffect effect = new WaterdropEffect();
        effect.setInfo(deadCell, culprit);
        return effect;
    }


    private class WaterdropEffect implements IEffect {
        private static final float FALL_ACCELERATION = 500.0f;
        private static final float FALL_VARIATION = 50.0f;
        private static final float COLOR_SPEED = 7.5f;
        private final float fallAcceleration;
        private Vector2 pos;
        private boolean dead;
        private Color cellColor;
        private Color dropColor;
        private float cellSize;
        private float fallSpeed;

        WaterdropEffect() {
            fallAcceleration = FALL_ACCELERATION + MathUtils.random(-FALL_VARIATION, FALL_VARIATION);
        }

        @Override
        public void setInfo(Cell deadCell, Vector2 culprit) {
            pos = deadCell.pos.cpy();
            cellSize = deadCell.size;
            cellColor = deadCell.getColorCopy();
            dropColor = new Color(cellColor.r, cellColor.g, cellColor.b, 0.0f);
        }

        @Override
        public void draw(Batch batch) {
            final float dt = Gdx.graphics.getDeltaTime();
            fallSpeed += fallAcceleration * dt;
            pos.y -= fallSpeed * dt;

            cellColor.set(
                    cellColor.r, cellColor.g, cellColor.b,
                    Math.max(cellColor.a - COLOR_SPEED * dt, 0.0f)
            );
            dropColor.set(
                    cellColor.r, cellColor.g, cellColor.b,
                    Math.min(dropColor.a + COLOR_SPEED * dt, 1.0f)
            );

            Cell.draw(cellColor, batch, pos.x, pos.y, cellSize);
            Cell.draw(dropTexture, dropColor, batch, pos.x, pos.y, cellSize);

            final Vector3 translation = batch.getTransformMatrix().getTranslation(new Vector3());
            dead = translation.y + pos.y + dropTexture.getHeight() < 0;
        }

        @Override
        public boolean isDone() {
            return dead;
        }
    }
}

File: Klooni1010/core/src/dev/lonami/klooni/effects/WaterdropEffectFactory.kt
package dev.lonami.klooni.effects

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.graphics.g2d.Batch
import com.badlogic.gdx.math.MathUtils
import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.math.Vector3
import dev.lonami.klooni.SkinLoader.loadPng
import dev.lonami.klooni.game.Cell
import dev.lonami.klooni.interfaces.Effect
import dev.lonami.klooni.interfaces.EffectFactory
import kotlin.math.max
import kotlin.math.min

class WaterdropEffectFactory : EffectFactory {

    private var dropTexture: Texture? = null
    override val name = "waterdrop"
    override val display = "Waterdrop"
    override val price = 200

    override fun create(deadCell: Cell, culprit: Vector2): Effect {
        if (dropTexture == null) dropTexture = loadPng("cells/drop.png")
        val effect: Effect = WaterdropEffect()
        effect.setInfo(deadCell, culprit)
        return effect
    }


    private inner class WaterdropEffect : Effect {
        private val fallAcceleration = FALL_ACCELERATION + MathUtils.random(-FALL_VARIATION, FALL_VARIATION)
        private var pos: Vector2? = null

        override var isDone: Boolean = false
            private set

        private var cellColor: Color? = null
        private var dropColor: Color? = null
        private var cellSize = 0f
        private var fallSpeed = 0f

        override fun setInfo(deadCell: Cell, culprit: Vector2) {
            pos = deadCell.pos.cpy()
            cellSize = deadCell.size
            cellColor = deadCell.colorCopy
            dropColor = Color(cellColor!!.r, cellColor!!.g, cellColor!!.b, 0.0f)
        }

        override fun draw(batch: Batch) {
            val dt = Gdx.graphics.deltaTime
            fallSpeed += fallAcceleration * dt
            pos!!.y -= fallSpeed * dt

            cellColor!!.set(
                cellColor!!.r, cellColor!!.g, cellColor!!.b,
                max(cellColor!!.a - COLOR_SPEED * dt, 0.0f)
            )
            dropColor!!.set(
                cellColor!!.r, cellColor!!.g, cellColor!!.b,
                min(dropColor!!.a + COLOR_SPEED * dt, 1.0f)
            )

            Cell.draw(cellColor, batch, pos!!.x, pos!!.y, cellSize)
            Cell.draw(dropTexture, dropColor, batch, pos!!.x, pos!!.y, cellSize)

            val translation = batch.transformMatrix.getTranslation(Vector3())
            this.isDone = translation.y + pos!!.y + dropTexture!!.height < 0
        }
    }

    companion object {
        private const val FALL_ACCELERATION = 500.0f
        private const val FALL_VARIATION = 50.0f
        private const val COLOR_SPEED = 7.5f
    }
}
